### PR TITLE
Add transform to control.ts and unit tests

### PR DESCRIPTION
**Pull request by neverbiasu**
This PR implemented 2 of css 2D transform-function(`rotate` and `translateX`), and also provided unit tests and integration tests.

### CODE REVIEW

File: src/living/helpers/babylonjs/InteractiveDynamicTexture.ts [hunk 1]
```diff
@@ -7,7 +7,8 @@ import { HTMLContentElement } from '../../nodes/HTMLContentElement';
 import { isHTMLContentElement } from '../../node-type';
 import { Control2D } from '../gui2d/control';
 import { domSymbolTree } from '../internal-constants';
-
+import DOMMatrixImpl from '../../geometry/DOMMatrix'
+import { postMultiply } from '../matrix-functions';
```
Comment (neverbiasu) [inline src/living/helpers/babylonjs/InteractiveDynamicTexture.ts:11]:
```suggestion
import { postMultiply } from '../matrix-functions';

```

File: src/living/helpers/gui2d/control.ts [hunk 2]
```diff
@@ -370,7 +392,7 @@ export class Control2D {
     const canvasContext = this._renderingContext;
     const boxRect = new DOMRectImpl(x, y, width, height);
     const hasTextChildren = this._isElementOwnsInnerText();
-
+
```
Comment (neverbiasu) [inline src/living/helpers/gui2d/control.ts:395]:
```suggestion
```

File: src/living/nodes/HTMLImageElement.ts [hunk 3]
```diff
@@ -135,6 +135,7 @@ export default class HTMLImageElementImpl extends HTMLContentElement implements
 
   private async _resizeImageData() {
     if (this._enableResizing && this._imageBitmap) {
+      console.log('height', this.height);
```
Comment (neverbiasu) [inline src/living/nodes/HTMLImageElement.ts:138]: remove comments

File: fixtures/css/css-transform.xsml [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+          const mouseY = event.y;
+          console.log('mouseY', mouseY);
+          const x = Math.round(mouseX);
+          img.style.transform = `translateX(${x - 512}px)`;
+          
+        });
+      </script>
+    </space>
+  </xsml>
+
```
Comment (neverbiasu) [inline fixtures/css/css-transform.xsml:72]:
```suggestion
 
```

File: fixtures/css/css-transform.xsml [hunk 5]
```diff
@@ -0,0 +1,72 @@
+<xsml>
+    <head>
+      <title>Spatial CSS Example (Basic)</title>
+      <style type="text/scss">
+        @material litered {
```
Comment (neverbiasu) [inline fixtures/css/css-transform.xsml:5]: remove

File: fixtures/css/css-transform.xsml [hunk 6]
```diff
@@ -0,0 +1,72 @@
...
+        #cube1 {
+          position: 1 1 1;
+          transform: rotate(10deg);
+          material: "liteblue";
+          animation: rotate 5s infinite;
+        }
+      </style>
+    </head>
+    <space>
+      <plane id="plane1">
```
Comment (neverbiasu) [inline fixtures/css/css-transform.xsml:33]:
```suggestion
      <plane> 
```

**Review by neverbiasu — COMMENTED** [review]
some format issues

File: src/living/helpers/gui2d/control.test.ts [hunk 7]
```diff
@@ -0,0 +1,41 @@
...
+    constructor() {
+        super(<any>{}, <any>{});
+
+        const mockContext: Partial<CanvasRenderingContext2D> = {
+            setTransform: (...args: any[]) => {
+                if (args.length === 6) {
+                    this.transform = new DOMMatrix(args);
+                } else if (args.length === 1 && typeof args[0] === 'object') {
+                    const transformInit: DOMMatrix2DInit = args[0];
+                    this.transform = new DOMMatrix([transformInit.m11, transformInit.m12, 0, 0,     transformInit.m21, transformInit.m22, 0, 0,     0, 0, 1, 0,     transformInit.m41, transformInit.m42, 0, 1]); // change m32 m33 to pass the test
```
Comment (neverbiasu) [inline src/living/helpers/gui2d/control.test.ts:16]:
```suggestion
                    this.transform = new DOMMatrix([
                      transformInit.m11, transformInit.m12, 0, 0,     
                      transformInit.m21, transformInit.m22, 0, 0,     
                      0, 0, 1, 0,     
                      transformInit.m41, transformInit.m42, 0, 1
                    ]); 
```

File: fixtures/css/css-transform.xsml [hunk 8]
```diff
@@ -0,0 +1,66 @@
...
+        const div = panel.querySelector('div');
+        const img = div.querySelector('img');
+        div.addEventListener('mousemove', (event) => {
+          const mouseX = event.x;
+          console.log('mouseX', mouseX);
+          const mouseY = event.y;
+          console.log('mouseY', mouseY);
+          const x = Math.round(mouseX);
+          img.style.transform = `translateX(${x - 512}px)`;
+
```
Comment (neverbiasu) [inline fixtures/css/css-transform.xsml:62]:
```suggestion
```

File: src/living/helpers/babylonjs/InteractiveDynamicTexture.ts [hunk 9]
```diff
@@ -337,6 +356,49 @@ export class InteractiveDynamicTexture extends BABYLON.DynamicTexture {
...
+        const x = parseFloat(transform.value);
+        const translateMatrix: DOMMatrix = new DOMMatrixImpl([
+          1, 0, 0, 0,  
+          0, 1, 0, 0,  
+          0, 0, 1, 0,  
+          x, 0, 0, 1
+        ]);
+        matrix = postMultiply(matrix, translateMatrix);
+      }
+
```
Comment (neverbiasu) [inline src/living/helpers/babylonjs/InteractiveDynamicTexture.ts:385]:
```suggestion
```

File: src/living/helpers/gui2d/control.test.ts [hunk 10]
```diff
@@ -0,0 +1,51 @@
...
+          }
+        },
+        getTransform: () => {
+          return this.transform;
+        },
+      };
+      this._renderingContext = mockContext as any;
+  }
+  prepareTransformMatrix(matrix: DOMMatrix) {
+      this.transform = matrix;
```
Comment (neverbiasu) [inline src/living/helpers/gui2d/control.test.ts:31]:
```suggestion
    this.transform = matrix;
```

File: src/living/helpers/gui2d/control.test.ts [hunk 11]
```diff
@@ -0,0 +1,51 @@
...
+  }
+  expect(matrix: DOMMatrix) {
+      expect(this._renderingContext.getTransform()).toEqual(matrix);
```
Comment (neverbiasu) [inline src/living/helpers/gui2d/control.test.ts:34]:
```suggestion
    expect(this._renderingContext.getTransform()).toEqual(matrix);
```

File: fixtures/css/css-transform.xsml [hunk 12]
```diff
@@ -0,0 +1,65 @@
...
+          }
+          img {
+            width: 25%;
+            height: 25%;
+            object-fit: cover;
+            position: relative;
+          }
+        </style>
+        <div>
+          <img src="./texture.jpg">
```
Comment (yorkie) [inline fixtures/css/css-transform.xsml:46]: The texture file should moved to fixtures/textures directory.

File: fixtures/css/css-transform.xsml [hunk 13]
```diff
@@ -0,0 +1,65 @@
...
+        </div>
+      </plane>
+      <cube class="m_cube" />
```
Comment (yorkie) [inline fixtures/css/css-transform.xsml:49]: What does this cube for? It seems not related with CSS transform :(

File: src/living/helpers/babylonjs/InteractiveDynamicTexture.test.ts [hunk 14]
```diff
@@ -0,0 +1,30 @@
+import { InteractiveDynamicTexture } from './InteractiveDynamicTexture';
+import DOMMatrix from '../../geometry/DOMMatrix'
```
Comment (yorkie) [inline src/living/helpers/babylonjs/InteractiveDynamicTexture.test.ts:2]:
```suggestion
import DOMMatrix from '../../geometry/DOMMatrix';
```

File: src/living/helpers/babylonjs/InteractiveDynamicTexture.test.ts [hunk 15]
```diff
@@ -0,0 +1,30 @@
...
+import { describe, it, expect } from '@jest/globals'
```
Comment (yorkie) [inline src/living/helpers/babylonjs/InteractiveDynamicTexture.test.ts:3]:
```suggestion
import { describe, it, expect } from '@jest/globals';
```

File: src/living/helpers/babylonjs/InteractiveDynamicTexture.test.ts [hunk 16]
```diff
@@ -0,0 +1,30 @@
...
+
+describe('InteractiveDynamicTexture', () => {
+  it('should return correct matrix', () => {
+    const transformStr = "rotate(90deg) translateX(10px)"
```
Comment (yorkie) [inline src/living/helpers/babylonjs/InteractiveDynamicTexture.test.ts:7]:
```suggestion
    const transformStr = 'rotate(90deg) translateX(10px)';
```

File: src/living/helpers/babylonjs/InteractiveDynamicTexture.test.ts [hunk 17]
```diff
@@ -0,0 +1,30 @@
...
+      -1, 0, 0, 0,  
+      0, 0, 1, 0,  
+      0, 10, 0, 1
+    ]);
+    const result = InteractiveDynamicTexture._parserTransform(transformStr);
+
+    expect(result).toEqual(expectedMatrix);
+  })
+  it('should return identity matrix when no transform is applied', () => {
+    const transformStr = " "
```
Comment (yorkie) [inline src/living/helpers/babylonjs/InteractiveDynamicTexture.test.ts:19]:
```suggestion
    const transformStr = ' ';
```

**Review by yorkie — CHANGES_REQUESTED** [review]

File: src/living/helpers/gui2d/control.ts [hunk 18]
```diff
@@ -164,12 +163,13 @@ export class Control2D {
   private _lastRect: DOMRectReadOnlyImpl;
   private _lastCursor: BABYLON.Vector2;
   private _isCursorInside = false;
-  private _renderingContext: CanvasRenderingContext2D;
+  protected _renderingContext: CanvasRenderingContext2D; // for unit test
   private _overwriteHeight: number;
   private _overwriteWidth: number;
-  private _imageData: ImageDataImpl;
+  private _imageData: ImageBitmap;
```
Comment (yorkie) [inline src/living/helpers/gui2d/control.ts:169]: 1. Private field should be a *impl type
2. The name should be `_imageBitmap` if its type is `ImageBitmap*`

File: src/living/helpers/gui2d/control.ts [hunk 19]
```diff
@@ -164,12 +163,13 @@ export class Control2D {
   private _lastRect: DOMRectReadOnlyImpl;
   private _lastCursor: BABYLON.Vector2;
   private _isCursorInside = false;
-  private _renderingContext: CanvasRenderingContext2D;
+  protected _renderingContext: CanvasRenderingContext2D; // for unit test
```
Comment (yorkie) [inline src/living/helpers/gui2d/control.ts:166]: Please remove the comments that is unrelated.

File: src/living/helpers/gui2d/control.ts [hunk 20]
```diff
@@ -164,12 +163,13 @@ export class Control2D {
...
   private _overwriteHeight: number;
   private _overwriteWidth: number;
-  private _imageData: ImageDataImpl;
+  private _imageData: ImageBitmap;
   private _isDirty = true;
-
+  private _transform: DOMMatrix;
+  private _currentTransform: DOMMatrix;
```
Comment (yorkie) [inline src/living/helpers/gui2d/control.ts:172]: Ditto, for the above 2 fields.

File: src/living/helpers/gui2d/control.ts [hunk 21]
```diff
@@ -197,7 +197,7 @@ export class Control2D {
     }
   }
 
-  setImageData(data: ImageDataImpl) {
+  setImageData(data: ImageBitmap) {
```
Comment (yorkie) [inline src/living/helpers/gui2d/control.ts:200]:
```suggestion
  setImageBitmap(bitmap: ImageBitmap) {
```

File: src/living/helpers/gui2d/control.ts [hunk 22]
```diff
@@ -739,9 +759,14 @@ export class Control2D {
       element.width = rect.width;
       element.height = rect.height;
     });
-    renderingContext.putImageData(this._imageData, rect.x, rect.y, 0, 0, rect.width, rect.height);
+    renderingContext.drawImage(this._imageData, rect.x, rect.y, rect.width, rect.height);
   }
-
+  
+  _updateTransform(renderingContext: CanvasRenderingContext2D = this._renderingContext) {
```
Comment (yorkie) [inline src/living/helpers/gui2d/control.ts:765]: No need to pass an argument of `renderingContext`, because it always be `this._renderingContext`, feel free to correct me if I'm wrong :)

File: src/living/helpers/gui2d/control.ts [hunk 23]
```diff
@@ -739,9 +759,14 @@ export class Control2D {
...
+    const transformMatrix = this.currentTransform;
```
Comment (yorkie) [inline src/living/helpers/gui2d/control.ts:766]: IMO it's better to keep the name consistent with the spec.

File: src/living/helpers/gui2d/control.ts [hunk 24]
```diff
@@ -389,6 +404,11 @@ export class Control2D {
       }
     }
 
+    /**
+     * Render the transform.
```
Comment (yorkie) [inline src/living/helpers/gui2d/control.ts:408]: This comment is incorrect, what this function does is just changing the transform matrix for the current rendering context IMO.

File: src/living/helpers/babylonjs/InteractiveDynamicTexture.ts [hunk 25]
```diff
@@ -313,6 +315,23 @@ export class InteractiveDynamicTexture extends BABYLON.DynamicTexture {
       isDirtyAfterRendering = control.isDirty();
     } else {
       layout = currentElementOrControl._control.layoutNode.getLayout();
+      const control = currentElementOrControl._control;
+      const style = control._style;
+      const transformStr = style.transform;
+      if (currentElementOrControl.parentElement === null) {
+        control.transform = InteractiveDynamicTexture._parserTransform(transformStr);
```
Comment (yorkie) [inline src/living/helpers/babylonjs/InteractiveDynamicTexture.ts:322]: This makes a performance issue when you parse the transform in every frame.

File: src/living/helpers/babylonjs/InteractiveDynamicTexture.ts [hunk 26]
```diff
@@ -313,6 +315,23 @@ export class InteractiveDynamicTexture extends BABYLON.DynamicTexture {
...
+      } else {
+        control.transform = InteractiveDynamicTexture._parserTransform(transformStr);
+        const parentElement = currentElementOrControl.parentElement;
+        if (parentElement instanceof HTMLContentElement) {
+          const parentControl = parentElement._control;
+          const parentTransform = parentControl.transform;
+          control.currentTransform = postMultiply(parentTransform, control.transform);
+          currentElementOrControl._control = control;
+        }
+      }
```
Comment (yorkie) [inline src/living/helpers/babylonjs/InteractiveDynamicTexture.ts:334]: The above snippet should not be here

File: src/living/helpers/babylonjs/InteractiveDynamicTexture.ts [hunk 27]
```diff
@@ -337,6 +356,48 @@ export class InteractiveDynamicTexture extends BABYLON.DynamicTexture {
     return isDirtyAfterRendering;
   }
 
+  // cannot call it in `render()`, cuz it cost too much.
+  static _parserTransform(transformStr: string): DOMMatrix {
+    const pattern: RegExp = /(translateX|rotate)\((\d+)(px|deg)\)/g;
```
Comment (yorkie) [inline src/living/helpers/babylonjs/InteractiveDynamicTexture.ts:361]: Is there any third-party library to parse transform?

File: src/living/helpers/babylonjs/InteractiveDynamicTexture.ts [hunk 28]
```diff
@@ -337,6 +356,48 @@ export class InteractiveDynamicTexture extends BABYLON.DynamicTexture {
     return isDirtyAfterRendering;
   }
 
+  // cannot call it in `render()`, cuz it cost too much.
+  static _parserTransform(transformStr: string): DOMMatrix {
```
Comment (yorkie) [inline src/living/helpers/babylonjs/InteractiveDynamicTexture.ts:360]: The CSS property parsing should not be a rendering-related source file, please do this work in CSS-related files.

File: src/living/helpers/babylonjs/InteractiveDynamicTexture.ts [hunk 29]
```diff
@@ -337,6 +356,48 @@ export class InteractiveDynamicTexture extends BABYLON.DynamicTexture {
     return isDirtyAfterRendering;
   }
 
+  // cannot call it in `render()`, cuz it cost too much.
```
Comment (yorkie) [inline src/living/helpers/babylonjs/InteractiveDynamicTexture.ts:359]: Actually you does call this function in `render()` :(

File: src/living/helpers/gui2d/control.test.ts [hunk 30]
```diff
@@ -0,0 +1,51 @@
...
+    };
+    this._renderingContext = mockContext as any;
+  }
+  prepareTransformMatrix(matrix: DOMMatrix) {
+    this.transform = matrix;
+  }
+  expect(matrix: DOMMatrix) {
+    expect(this._renderingContext.getTransform()).toEqual(matrix);
+  }
+}
```
Comment (yorkie) [inline src/living/helpers/gui2d/control.test.ts:36]: I think there is no necessary to create a new class to do this work, it looks too complex to me.

File: src/living/helpers/gui2d/control.test.ts [hunk 31]
```diff
@@ -0,0 +1,51 @@
...
+    const control = new MockControl();
+    const matrix = new DOMMatrix([
+        0, 1, 0, 0,    
+        -1, 0, 0, 0,    
+        0, 0, 1, 0,    
+        10, 0, 0, 1
+    ]);
+    control.prepareTransformMatrix(matrix);
+    control._updateTransform();
+    control.expect(matrix);
```
Comment (yorkie) [inline src/living/helpers/gui2d/control.test.ts:49]: Mock class should do mocking only, expectation is not the scope of mock.

File: src/living/helpers/babylonjs/InteractiveDynamicTexture.test.ts [hunk 32]
```diff
@@ -0,0 +1,30 @@
...
+describe('InteractiveDynamicTexture', () => {
+  it('should return correct matrix', () => {
+    const transformStr = 'rotate(90deg) translateX(10px)';
+    const expectedMatrix = new DOMMatrix([
+      0, 1, 0, 0,   
+      -1, 0, 0, 0,  
+      0, 0, 1, 0,  
+      0, 10, 0, 1
+    ]);
+    const result = InteractiveDynamicTexture._parserTransform(transformStr);
```
Comment (yorkie) [inline src/living/helpers/babylonjs/InteractiveDynamicTexture.test.ts:14]: Ditto, `InteractiveDynamicTexture` should not contain a function that does parsing, you have to finish it in CSS parsing stage in this project.

**Review by yorkie — CHANGES_REQUESTED** [review]

File: src/living/helpers/gui2d/control.test.ts (hunk 30 above)
Comment (neverbiasu) [reply src/living/helpers/gui2d/control.test.ts:33]: we need to do it cuz the Control2D need 2 element which we cannot provide.

File: src/living/helpers/gui2d/control.ts [hunk 33]
```diff
@@ -739,9 +763,77 @@ export class Control2D {
...
+        ]);
+        transformMatrix = postMultiply(transformMatrix, rotateMatrix) as DOMMatrixImpl;
+      }
+    });
+    return transformMatrix;
   }
 
+  _updateTransform() {
+    const renderingContext = this._renderingContext;
+    const transformMatrix = this.currentTransformMatrix;
```
Comment (neverbiasu) [inline src/living/helpers/gui2d/control.ts:833]:
```suggestion
    const currentTransformMatrix = this.currentTransformMatrix;
```

File: src/living/helpers/gui2d/control.ts [hunk 34]
```diff
@@ -739,9 +763,77 @@ export class Control2D {
...
+    renderingContext.setTransform(transformMatrix);
```
Comment (neverbiasu) [inline src/living/helpers/gui2d/control.ts:834]:
```suggestion
    renderingContext.setTransform(currentTransformMatrix);
```

File: src/living/helpers/gui2d/control.ts [hunk 35]
```diff
@@ -739,9 +763,77 @@ export class Control2D {
...
+          const parentControl = parentElement._control;
+          const parentTransform = parentControl.transformMatrix;
+          this.currentTransformMatrix = postMultiply(parentTransform, this.transformMatrix);
+          element._control = this;
+        }
+      }
+    }
+  }
+  
+  static _parserTransform(transformStr: string): DOMMatrixImpl {
```
Comment (yorkie) [inline src/living/helpers/gui2d/control.ts:790]: CSS related parsing functions should be moved to `cssom`, please.

**Review by yorkie — CHANGES_REQUESTED** [review]

File: src/living/cssom/parsers/index.ts [hunk 36]
```diff
@@ -950,3 +952,44 @@ export function shorthandSetter(
     }
   };
 }
+
+export function parserTransform(transformStr: string): DOMMatrixImpl {
```
Comment (yorkie) [inline src/living/cssom/parsers/index.ts:956]:
```suggestion
export function parseTransform(transformStr: string): DOMMatrixImpl {
```

And a parse function should not return a DOMMatrix

File: src/living/cssom/parsers/index.ts [hunk 37]
```diff
@@ -950,3 +952,21 @@ export function shorthandSetter(
     }
   };
 }
+
+export class Transform {
+  type: string;
+  value: string;
+  unit: string;
```
Comment (yorkie) [inline src/living/cssom/parsers/index.ts:959]:
```suggestion
  unit: 'px' | 'deg';
```

File: src/living/cssom/parsers/index.ts [hunk 38]
```diff
@@ -3,6 +3,8 @@ import { hslToRgb } from '../utils/color-space';
 import { colorNames, namedColors } from './named-colors';
 import { defineSpatialProperty } from '../spatial-properties/helper';
 import * as animationShorthandParser from './animation-shorthand/index';
+import DOMMatrixImpl from '../../geometry/DOMMatrix';
+import { postMultiply } from '../../helpers/matrix-functions';
 
 const integerRegEx = /^[-+]?[0-9]+$/;
 const numberRegEx = /^[-+]?[0-9]*\.?[0-9]+$/;
```
Comment (yorkie) [inline src/living/cssom/parsers/index.ts:10]: Your RegExp should be defined here.

File: src/living/helpers/gui2d/control.ts [hunk 39]
```diff
@@ -739,9 +756,71 @@ export class Control2D {
...
+  private _updateCurrentTransformMatrix() {
+    const element = this._element;
+    const style = this._style;
+    if (element instanceof ShadowRootImpl) {
+      return;
+    } 
+    const transformStr = style.transform;
+    const parentElement = element.parentElement;
+    const transforms = parseTransform(transformStr)
+    this.currentTransformMatrix = this._calculateTransformMatrix(transforms);
```
Comment (yorkie) [inline src/living/helpers/gui2d/control.ts:771]: You could move this in `currentTransformMatrix ` setter.

File: src/living/helpers/gui2d/control.ts [hunk 40]
```diff
@@ -739,9 +756,71 @@ export class Control2D {
...
+    if (parentElement === null) {
```
Comment (yorkie) [inline src/living/helpers/gui2d/control.ts:772]:
```suggestion
    if (parentElement == null) {
```
Use non-strict equal for null check.

File: src/living/helpers/gui2d/control.ts [hunk 41]
```diff
@@ -739,9 +756,71 @@ export class Control2D {
...
+      if (!isHTMLContentElement(parentElement)) {
+        return;
+      } 
+      const parentControl = parentElement._control;
+      const parentCTM = parentControl.currentTransformMatrix;
+      this.currentTransformMatrix = postMultiply(parentCTM, this.currentTransformMatrix);
+    }
+  }
+  
+  _calculateTransformMatrix(transforms: { type: string, value: string, unit: string }[]): DOMMatrixImpl {
```
Comment (yorkie) [inline src/living/helpers/gui2d/control.ts:784]:
```suggestion
  _calculateTransformMatrix(transforms: Transform[]): DOMMatrixImpl {
```

File: src/living/helpers/gui2d/control.ts [hunk 42]
```diff
@@ -739,9 +756,71 @@ export class Control2D {
...
 
+  private _updateCurrentTransformMatrix() {
+    const element = this._element;
+    const style = this._style;
+    if (element instanceof ShadowRootImpl) {
+      return;
+    } 
+    const transformStr = style.transform;
+    const parentElement = element.parentElement;
+    const transforms = parseTransform(transformStr)
```
Comment (yorkie) [inline src/living/helpers/gui2d/control.ts:770]:
```suggestion
    const transforms = parseTransform(transformStr);
```

**Review by yorkie — CHANGES_REQUESTED** [review]

File: src/living/cssom/parsers/index.ts [hunk 43]
```diff
@@ -950,3 +951,87 @@ export function shorthandSetter(
...
+  }
+}
+
+export class RotationTransformFunction extends TransformFunction<PropertyAngleValue> {
+  constructor(name: string, args: string[]) {
+    const values = processArgs(args, toAngleStr) as PropertyAngleValue[];
+    super(name, values);
+  }
+  get isValid(): boolean {
+    return this.values.length > 0;
```
Comment (yorkie) [inline src/living/cssom/parsers/index.ts:1012]:
```suggestion
    return this.values.length === 1;
```

See: https://developer.mozilla.org/en-US/docs/Web/CSS/transform-function/rotate

File: src/living/cssom/parsers/index.ts [hunk 44]
```diff
@@ -950,3 +951,87 @@ export function shorthandSetter(
...
+  }
+}
+
+export type UnionTransformFunction = TranslationTransformFunction | RotationTransformFunction;
+
+export function parseTransform(transformStr: string): UnionTransformFunction[] {
+  let parsedTransformFunctions: UnionTransformFunction[] = [];
```
Comment (yorkie) [inline src/living/cssom/parsers/index.ts:1019]:
```suggestion
  const parsedTransformFunctions: UnionTransformFunction[] = [];
```

File: src/living/cssom/parsers/index.ts [hunk 45]
```diff
@@ -950,3 +951,87 @@ export function shorthandSetter(
...
+  let parsedResult: string[] = [];
+  while ((parsedResult = transformFunctionRegEx.exec(transformStr)) !== null) {
+    const transformFunctionName: string = parsedResult[1];
+    const transformFunctionArgs: string[] = parsedResult[2].split(',').map(param => param.trim());
```
Comment (yorkie) [inline src/living/cssom/parsers/index.ts:1023]:
```suggestion
    const transformFunctionArgs: string[] = parsedResult[2].split(',').map(arg => arg.trim());
```

File: src/living/cssom/parsers/index.ts [hunk 46]
```diff
@@ -950,3 +951,87 @@ export function shorthandSetter(
...
+    if (transformFunctionName === 'rotate') {
+      if (!createAndAddTransformFunction(RotationTransformFunction, transformFunctionName, transformFunctionArgs, parsedTransformFunctions)) {
```
Comment (yorkie) [inline src/living/cssom/parsers/index.ts:1025]:
```suggestion
      if (!createAndAddTransformFunctionTo(parsedTransformFunctions, {
        constructor: RotationTransformFunction, name: transformFunctionName, args: transformFunctionArgs
      })) {
```

File: src/living/helpers/gui2d/control.ts [hunk 47]
```diff
@@ -739,9 +756,57 @@ export class Control2D {
...
+        transformMatrix = translate(transformMatrix, x, 0, 0);
+      } else if (transformFunction.name === 'rotate') {
+        const angle = transformFunction.values[0].value as number;
+        transformMatrix = rotate(transformMatrix, angle);
+      }
+    });
+    return transformMatrix;
   }
 
+  updateCtxTransform() {
```
Comment (yorkie) [inline src/living/helpers/gui2d/control.ts:804]:
```suggestion
  syncRenderingContextTransform() {
```

File: src/living/helpers/gui2d/control.ts [hunk 48]
```diff
@@ -739,9 +756,57 @@ export class Control2D {
...
+    let transformMatrix = new DOMMatrixImpl([
+      1, 0, 0, 0,   
+      0, 1, 0, 0,  
+      0, 0, 1, 0,   
+      0, 0, 0, 1
+    ]);
+    transformFunctions.forEach(transformFunction => {
+      if (transformFunction.name === 'translateX') {
+        const x = transformFunction.values[0].value['number'];
+        transformMatrix = translate(transformMatrix, x, 0, 0);
```
Comment (yorkie) [inline src/living/helpers/gui2d/control.ts:795]:
```suggestion
        transformMatrix = translate(transformMatrix, transformFunction.x, transformFunction.y, transformFunction.z);
```

File: src/living/helpers/gui2d/control.ts [hunk 49]
```diff
@@ -739,9 +756,57 @@ export class Control2D {
...
+      } else if (transformFunction.name === 'rotate') {
+        const angle = transformFunction.values[0].value as number;
+        transformMatrix = rotate(transformMatrix, angle);
```
Comment (yorkie) [inline src/living/helpers/gui2d/control.ts:798]:
```suggestion
        transformMatrix = rotate(transformMatrix, transformFunction.angle);
```

File: src/living/helpers/matrix-functions.ts [hunk 50]
```diff
@@ -15,7 +15,7 @@ export function postMultiply(self: DOMMatrix, other: DOMMatrix): DOMMatrix {
   return new DOMMatrixImpl(resElements);
 }
 
-export function preMultiply(other: DOMMatrix, self: DOMMatrix): DOMMatrix { 
+export function preMultiply(other: DOMMatrix, self: DOMMatrix): DOMMatrixImpl {
```
Comment (yorkie) [inline src/living/helpers/matrix-functions.ts:18]: Why does this PR change the returned value type?

File: src/living/helpers/transform-functions.ts [hunk 51]
```diff
@@ -0,0 +1,24 @@
+import DOMMatrixImpl from '../geometry/DOMMatrix';
```
Comment (yorkie) [inline src/living/helpers/transform-functions.ts:1]: These 2 functions seems to be moved in `matrix-functions.ts`.

Comment (yorkie) [reply src/living/helpers/transform-functions.ts:1]: Or change this source filename to "matrix-transform-functions.ts".

File: src/living/cssom/parsers/index.test.ts [hunk 52]
```diff
@@ -380,3 +380,37 @@ describe('implicitSetter', () => {
...
+    expect(result).toEqual([
+      new parsers.TranslationTransformFunction('translateX', ['10px']),
+      new parsers.RotationTransformFunction('rotate', ['45deg'])
+    ]);
+  });
+
+  it('should process input function name error correctly', () => {
+    const result = parsers.parseTransform('translateM(10px)');
+    expect(result).toEqual([]);
+  })
```
Comment (yorkie) [inline src/living/cssom/parsers/index.test.ts:410]:
```suggestion
  });
```

File: src/living/cssom/parsers/index.test.ts [hunk 53]
```diff
@@ -380,3 +380,37 @@ describe('implicitSetter', () => {
...
+
+  it('should process input values error correctly', () => {
+    const result = parsers.parseTransform('translateX(10py)');
+    expect(result).toEqual([]);
+  })
```
Comment (yorkie) [inline src/living/cssom/parsers/index.test.ts:415]:
```suggestion
  });
```

File: src/living/cssom/parsers/index.ts [hunk 54]
```diff
@@ -18,6 +18,7 @@ const calcRegEx = /^calc\(([^)]*)\)$/;
 const colorRegEx4 =
   /^hsla?\(\s*(-?\d+|-?\d*.\d+)\s*,\s*(-?\d+|-?\d*.\d+)%\s*,\s*(-?\d+|-?\d*.\d+)%\s*(,\s*(-?\d+|-?\d*.\d+)\s*)?\)/;
 const angleRegEx = /^([-+]?[0-9]*\.?[0-9]+)(deg|grad|rad)$/;
+const transformFunctionRegEx = /(\w+)\((\w+)\)/g;
```
Comment (yorkie) [inline src/living/cssom/parsers/index.ts:21]:
```suggestion
const css3TransformFunctionNames = [
  'translate',
  'translateX',
  'translateY',
  ...
];
const transformFunctionRegEx = new RegExp(`(${css3TransformFunctionNames.join('|')})\(\w+\)`, 'g');
```

File: src/living/cssom/parsers/index.ts [hunk 55]
```diff
@@ -950,3 +951,87 @@ export function shorthandSetter(
...
+
+  constructor(name: string, values: Tv[] = []) {
+    this.name = name;
+    this.values = values;
+  }
+}
+
+export class TranslationTransformFunction extends TransformFunction<PropertyLengthValue> {
+  constructor(name: string, args: string[]) {
+    const values = processArgs(args, toLengthStr) as PropertyLengthValue[];
```
Comment (yorkie) [inline src/living/cssom/parsers/index.ts:998]:
```suggestion
    const values = super.processArgs(args, toLengthStr);
```

File: src/living/helpers/gui2d/control.ts [hunk 56]
```diff
@@ -739,9 +756,57 @@ export class Control2D {
...
+        return;
+      } 
+      const parentControl = parentElement._control;
+      const parentCTM = parentControl.currentTransformMatrix;
+      this.currentTransformMatrix = postMultiply(parentCTM, transformMatrix);
+    }
+  }
+  
+  calculateTransformMatrix(transformFunctions: UnionTransformFunction[]): DOMMatrixImpl {
+    let transformMatrix = new DOMMatrixImpl([
```
Comment (yorkie) [inline src/living/helpers/gui2d/control.ts:786]:
```suggestion
    let transformMatrix: DOMMatrix = new DOMMatrixImpl([
```

**Review by yorkie — CHANGES_REQUESTED** [review]

File: src/living/helpers/gui2d/control.ts [hunk 57]
```diff
@@ -739,9 +755,56 @@ export class Control2D {
       element.width = rect.width;
       element.height = rect.height;
     });
-    renderingContext.putImageData(this._imageData, rect.x, rect.y, 0, 0, rect.width, rect.height);
+    /**
+     * NOTE|feichi|: `putImageData` paints image onto canvas without transform,
```
Comment (yorkie) [inline src/living/helpers/gui2d/control.ts:759]:
```suggestion
     * NOTE(feichi): `putImageData` paints image onto canvas without transform,
```

File: src/living/helpers/gui2d/control.ts [hunk 58]
```diff
@@ -739,9 +755,60 @@ export class Control2D {
...
+  private _updateCurrentTransformMatrix() {
+    const element = this._element;
+    if (element instanceof ShadowRootImpl) {
+      return;
+    } 
+    const transformStr = this._style.transform;
+    const parentElement = element.parentElement;
+    const transformMatrix = this.calculateTransformMatrix(parseTransform(transformStr));
+    /**
+     * If the parentElement is the direct childNodes of ShadowRoot, it will be null.
```
Comment (neverbiasu) [inline src/living/helpers/gui2d/control.ts:775]: add an example to explain

File: src/living/helpers/gui2d/control.ts [hunk 59]
```diff
@@ -739,9 +755,60 @@ export class Control2D {
       element.width = rect.width;
       element.height = rect.height;
     });
-    renderingContext.putImageData(this._imageData, rect.x, rect.y, 0, 0, rect.width, rect.height);
+    /**
+     * NOTE(feichi): `putImageData` paints image onto canvas without transform,
```
Comment (neverbiasu) [inline src/living/helpers/gui2d/control.ts:759]: not clear enough

**Review by yorkie — APPROVED** [review]
LGTM
